Made improvements to takeSample. Also changed SparkLocalKMeans to SparkKMeans

edisontung opened the pull request:
Instead of calling count on an RDD multiple times, stored the count so it can be used. Also, called the collect method for the array earlier also to reduce the number of times count() is called.

edisontung committed:
Added KMeans examples

LocalKMeans runs locally with a randomly generated dataset.
SparkLocalKMeans takes an input file and runs KMeans on it.

edisontung committed:
added takeSamples method

takeSamples method takes a specified number of samples from the RDD and
outputs it in an array.

edisontung committed:
Renamed SparkLocalKMeans to SparkKMeans

edisontung committed:
Fixed bug in RDD

Math.min takes 2 args, not 1. This was not committed earlier for some
reason

edisontung committed:
Revert de01b6deaaee1b43321e0aac330f4a98c0ea61c6^..HEAD

mateiz commented on core/src/main/scala/spark/RDD.scala:
```diff
@@ -91,6 +91,41 @@ abstract class RDD[T: ClassManifest](@transient sc: SparkContext) extends Serial
+    var samples = r.collect()
+	
+  	while (samples.length < total) {
+  		r = new SampledRDD(this, withReplacement, fraction, seed)
```
Here you are not calling samples = r.collect() again, so samples.length will always remain less than total.

mateiz commented on core/src/main/scala/spark/RDD.scala:
```diff
@@ -91,6 +91,41 @@ abstract class RDD[T: ClassManifest](@transient sc: SparkContext) extends Serial
+	
+  	for (i <- 0 to total - 1) {
+  		arr(i) = samples(i)
+  	}
```
Instead of this you could write val arr = samples.take(total), or just return samples.take(total).

mateiz commented on core/src/main/scala/spark/RDD.scala:
```diff
@@ -91,6 +91,41 @@ abstract class RDD[T: ClassManifest](@transient sc: SparkContext) extends Serial
+  		fraction = 1.0
+  	}
+  	else if (num < 0) {
+  		throw(new IllegalArgumentException())
```
Add some text in the exception ("Negative number of elements requested")

mateiz commented on core/src/main/scala/spark/RDD.scala:
```diff
@@ -91,6 +91,41 @@ abstract class RDD[T: ClassManifest](@transient sc: SparkContext) extends Serial
+  		throw(new IllegalArgumentException())
+  	}
+  	else {
+  		fraction = Math.min(multiplier*(num+1)/count(), 1.0)
```
Here you are still calling count() instead of using initialCount.

mateiz commented on core/src/main/scala/spark/RDD.scala:
```diff
@@ -91,6 +91,41 @@ abstract class RDD[T: ClassManifest](@transient sc: SparkContext) extends Serial
+  	}
+	
+  	var r = new SampledRDD(this, withReplacement, fraction, seed)
+    var samples = r.collect()
```
You could probably also just say val samples = this.sample(withReplacement, fraction, seed).collect()

mateiz commented:
Hey Edison,

This looks good but I made a few comments on the diff, including finding a couple of bugs (another extra call to count() and a potential infinite loop).

The other thing I noticed is that your new kmeans_data.txt is huge -- 750K lines. It would be better to replace it with the simple one we had before (or just generate another small one), so that people can look at it more easily.

edisontung committed:
Fixed bugs

I've fixed the bugs detailed in the diff. One of the bugs was already
fixed on the local file (forgot to commit).

mateiz commented:
Looks good. Thanks!
